File: prepare/s1/src/w1/p06/Solution.java
package w1.p06;
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans = new int[nums.length];
        int[] l = new int[nums.length];
        int[] r = new int[nums.length];
        for( int i=0; i<nums.length; i++){
            l[i] = (i==0) ? 1 : l[i-1] * nums[i-1];
        }
        for(int i=nums.length-1; i>=0; i--){
            r[i] = (i==nums.length-1) ? 1 : r[i+1] * nums[i+1];
        }
        
        for(int i=0; i<nums.length; i++){
            ans[i] = l[i]*r[i];
        }
        return ans;
    }
}

File: prepare/s1/src/w1/p06/Solution.py
from typing import List
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        l = [1] * n #left array
        r = [1] * n #right array

        for i in range(1, n):
            l[i] = l[i-1]*nums[i-1] 
        for i in range(n-2, -1, -1):
            r[i] = r[i+1]*nums[i+1] 

        return [ ( l[i] * r[i] ) for i in range(0,n) ]

        